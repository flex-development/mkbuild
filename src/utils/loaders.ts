/**
 * @file Utilities - loaders
 * @module mkbuild/utils/loaders
 */

import type * as pathe from '@flex-development/pathe'
import type * as esbuild from 'esbuild'

/**
 * Returns an esbuild [build api][1] [`loader`][2] configuration.
 *
 * [1]: https://esbuild.github.io/api/#build-api
 * [2]: https://esbuild.github.io/api/#loader
 *
 * @param {esbuild.Format} [format='esm'] - Output file format
 * @param {boolean} [bundle=false] - Bundling enabled?
 * @return {Record<pathe.Ext, esbuild.Loader>} `loader` configuration
 */
const loaders = (
  format: esbuild.Format = 'esm',
  bundle: boolean = false
): Record<pathe.Ext, esbuild.Loader> => ({
  '.cjs': format === 'cjs' && !bundle ? 'copy' : 'js',
  '.css': bundle ? 'css' : 'copy',
  '.cts': 'ts',
  '.d.cts': 'copy',
  '.d.mts': 'copy',
  '.d.ts': 'copy',
  '.data': bundle ? 'binary' : 'copy',
  '.eot': 'copy',
  '.gif': 'copy',
  '.jpeg': 'copy',
  '.jpg': 'copy',
  '.js': 'js',
  '.json': bundle ? 'json' : 'copy',
  '.json5': bundle ? 'json' : 'copy',
  '.jsonc': bundle ? 'json' : 'copy',
  '.jsx': 'jsx',
  '.mjs': format === 'esm' && !bundle ? 'copy' : 'js',
  '.mts': 'ts',
  '.otf': 'copy',
  '.png': 'copy',
  '.svg': 'copy',
  '.ts': 'ts',
  '.tsx': 'tsx',
  '.txt': bundle ? 'text' : 'copy',
  '.woff': 'copy',
  '.woff2': 'copy'
})

export default loaders
