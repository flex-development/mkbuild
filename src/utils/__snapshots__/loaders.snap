// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/loaders > cjs > should return loader configuration for cjs bundle 1`] = `
{
  ".cjs": "js",
  ".css": "css",
  ".cts": "ts",
  ".d.cts": "copy",
  ".d.mts": "copy",
  ".d.ts": "copy",
  ".data": "binary",
  ".eot": "copy",
  ".gif": "copy",
  ".jpeg": "copy",
  ".jpg": "copy",
  ".js": "js",
  ".json": "json",
  ".json5": "json",
  ".jsonc": "json",
  ".jsx": "jsx",
  ".mjs": "js",
  ".mts": "ts",
  ".otf": "copy",
  ".png": "copy",
  ".svg": "copy",
  ".ts": "ts",
  ".tsx": "tsx",
  ".txt": "text",
  ".woff": "copy",
  ".woff2": "copy",
}
`;

exports[`unit:utils/loaders > cjs > should return loader configuration for cjs transpilation 1`] = `
{
  ".cjs": "copy",
  ".css": "copy",
  ".cts": "ts",
  ".d.cts": "copy",
  ".d.mts": "copy",
  ".d.ts": "copy",
  ".data": "copy",
  ".eot": "copy",
  ".gif": "copy",
  ".jpeg": "copy",
  ".jpg": "copy",
  ".js": "js",
  ".json": "copy",
  ".json5": "copy",
  ".jsonc": "copy",
  ".jsx": "jsx",
  ".mjs": "js",
  ".mts": "ts",
  ".otf": "copy",
  ".png": "copy",
  ".svg": "copy",
  ".ts": "ts",
  ".tsx": "tsx",
  ".txt": "copy",
  ".woff": "copy",
  ".woff2": "copy",
}
`;

exports[`unit:utils/loaders > esm > should return loader configuration for esm bundle 1`] = `
{
  ".cjs": "js",
  ".css": "css",
  ".cts": "ts",
  ".d.cts": "copy",
  ".d.mts": "copy",
  ".d.ts": "copy",
  ".data": "binary",
  ".eot": "copy",
  ".gif": "copy",
  ".jpeg": "copy",
  ".jpg": "copy",
  ".js": "js",
  ".json": "json",
  ".json5": "json",
  ".jsonc": "json",
  ".jsx": "jsx",
  ".mjs": "js",
  ".mts": "ts",
  ".otf": "copy",
  ".png": "copy",
  ".svg": "copy",
  ".ts": "ts",
  ".tsx": "tsx",
  ".txt": "text",
  ".woff": "copy",
  ".woff2": "copy",
}
`;

exports[`unit:utils/loaders > esm > should return loader configuration for esm transpilation 1`] = `
{
  ".cjs": "js",
  ".css": "copy",
  ".cts": "ts",
  ".d.cts": "copy",
  ".d.mts": "copy",
  ".d.ts": "copy",
  ".data": "copy",
  ".eot": "copy",
  ".gif": "copy",
  ".jpeg": "copy",
  ".jpg": "copy",
  ".js": "js",
  ".json": "copy",
  ".json5": "copy",
  ".jsonc": "copy",
  ".jsx": "jsx",
  ".mjs": "copy",
  ".mts": "ts",
  ".otf": "copy",
  ".png": "copy",
  ".svg": "copy",
  ".ts": "ts",
  ".tsx": "tsx",
  ".txt": "copy",
  ".woff": "copy",
  ".woff2": "copy",
}
`;
